### DISCUSSION TITLE
Relais Ansteuerung

### DISCUSSION THREAD

**Discussion by Fillip38** [frickelzeugs/FingerprintDoorbell#6] [Ideas]
Hallo,
Ich bin natürlich auch über das YouTube Video auf das coole Projekt aufmerksam geworden und hab mir gleich mal den Fingersensor bestellt. Die Firmware habe ich schon mal probeweise auf dem ESP Installiert.
Ich hätte nur eine "Idee":
Wäre es denn möglich bzw kannst du dort noch die Möglichkeit einbauen, Relais anzusteuern, sodass man dort z.b die Klingel oder einen Summer ansteuern kann (von extern, MQTT bsp.) ?

**Comment by henfri** [comment]
Hello,

this could be a simple solution:
```
    case ScanResult::matchFound:
      notifyClients( String("Match Found: ") + match.matchId + " - " + match.matchName  + " with confidence of " + match.matchConfidence );
      if (match.scanResult != lastMatch.scanResult) {
        if (checkPairingValid()) {
          mqttClient.publish((String(mqttRootTopic) + "/ring").c_str(), "off");
          mqttClient.publish((String(mqttRootTopic) + "/matchId").c_str(), String(match.matchId).c_str());
          mqttClient.publish((String(mqttRootTopic) + "/matchName").c_str(), match.matchName.c_str());
          mqttClient.publish((String(mqttRootTopic) + "/matchConfidence").c_str(), String(match.matchConfidence).c_str());
          Serial.println("MQTT message sent: Open the door!");
          if (strstr(sent, match.MatchName) != NULL){
                 digitalWrite(customOutput1, HIGH); 
                 delay(500)
                 digitalWrite(customOutput1, LOW); 
                 Serial.println("Triggering Custom GPIO");
          }
        } else {
          notifyClients("Security issue! Match was not sent by MQTT because of invalid sensor pairing! This could potentially be an attack! If the sensor is new or has been replaced by you do a (re)pairing in settings page.");
        }
      }
```

**Reply by Fillip38** [reply]
Yes, that would work, but I would like to initiate the command to open externally (in my case via FHEM) because I want to control the door opening over it

**Reply by frickelzeugs** [reply]
Hi, I'm on vacation this week and have no access to a PC, but the feature you mentioned I've already implemented for a friend, but its undocumented and also disabled by default. You have to set the  "CUSTOM_GPIOS" pre processor define and make you own build to enable it. Also look in the code for how to use the 2 custom output GPIOs.

**Reply by henfri** [reply]
Yes, but you have lots of fans, and at least two of them want the Microcontroller to directly open the door.
Would be annoying if FHEM is down and the opening does not work.

Would you be willing to integrate this functionality?

Apart from that: Could you document the "customOutput"?

**Reply by henfri** [reply]
Ah, our posts overlapped :-) 

I have looked into the customOutput and Input. I could document it, if you accept Pull requests on this.
I would like to extend this, to also accept "1" and "0" instead of on/off. What do you think?

**Reply by frickelzeugs** [reply]
Sure, go ahead! But I can review/merge it at earliest next week after my vacation. Maybe I should also enable this per default then, if more of you guys find this useful.

**Comment by baetzmr** [comment]
Hello. One short question about the doorbellOutputPin = 19. is this GPIO Output high or low at startup? I want to build a Version using a Relay connected to this Output and want to avoid a ringing Doorbell when the ESP is rebooting 

PS: Still waiting for Sensor but preparations are ongoing ;-)

![image](https://user-images.githubusercontent.com/81546284/201782073-6c7f382b-3b68-4939-a64d-48fcc9466c60.jpeg)

**Reply by frickelzeugs** [reply]
Hi, GPIO19 will be low at startup and the bell does not ring ;-)

**Reply by baetzmr** [reply]
Danke sehr. Hätte ich ja auch dem Code entnehmen können. Aber danke für die Info. Sensor kam heute an ;-) Vielleicht schaffe ich heute im Laufe des Abends noch eine Testschaltung nach dem alles gelötet ist. Rückinfo kommt ob dann alles passt inklusive ioBroker Einbindung.

**Comment by BadPoldi** [comment]
Hello,

is there a documentation about the Custom GPIO how to enable/use it?

#ifdef CUSTOM_GPIOS
  const int   customOutput1 = 18; // not used internally, but can be set over MQTT
  const int   customOutput2 = 26; // not used internally, but can be set over MQTT
  const int   customInput1 = 21; // not used internally, but changes are published over MQTT
  const int   customInput2 = 22; // not used internally, but changes are published over MQTT
  bool customInput1Value = false;
  bool customInput2Value = false;
#endif

Regards,
